validphys2/src/validphys/plotutils.py

```diff
@@ -13,45 +13,120 @@
 import scipy.stats as stats
 
 import numpy as np
-import matplotlib.pyplot as plt
+from matplotlib.figure import Figure
+import matplotlib as mpl
 import matplotlib.scale as mscale
 import matplotlib.patches as mpatches
 import matplotlib.collections as mcollections
-from matplotlib  import transforms
+from matplotlib import transforms
 from matplotlib.markers import MarkerStyle
 from matplotlib import ticker
 
 from reportengine.floatformatting import format_number
 
 log = logging.getLogger(__name__)
 
+
+def subplots(figsize=(6.4, 4.8), nrows=1, ncols=1, sharex=False, sharey=False, **kwargs):
+    """
+    matplotlib.figure wrapper used to generate subplots.
+
+    Use matplotlib.figure.Figure() objects to avoid importing ``pyplot`` anywhere.
+    The reason is that ``pyplot`` maintains a global state that makes it misbehave
+    in multithreaded applications such when executed under dask parallel mode.
+
+    Parameters
+    ----------
+    figsize : 2-tuple of floats
+            defaults is (6.4, 4.8), that is, the same that would be set by mpl.rcParams["figure.figsize"]
+            avoids global figsize state set by rcParams
+    nrows, ncols : int, default 1
+
+    sharex, sharey : bool, default False
+
+    Returns
+    -------
+    tuple
+        fig, ax = (matplotlib.figure.Figure, fig.subplots)
+    """
+    fig = Figure(figsize=figsize)
+    ax = fig.subplots(nrows=nrows, ncols=ncols, sharex=sharex, sharey=sharey, **kwargs)
+    return fig, ax
+
+
+def add_subplot(figsize=(6.4, 4.8), projection=None, **kwargs):
+    """
+    matplotlib.figure wrapper used to add subplots.
+
+    Use matplotlib.figure.Figure() objects to avoid importing ``pyplot`` anywhere.
+    The reason is that ``pyplot`` maintains a global state that makes it misbehave
+    in multithreaded applications such when executed under dask parallel mode.
+
+    Parameters
+    ----------
+    figsize : 2-tuple of floats
+            default is (6.4, 4.8), that is, the same that would be set by mpl.rcParams["figure.figsize"]
+            avoids global figsize state set by rcParams
+    projections : The projection type of the subplot (Axes).
+                default is None
+
+    Returns
+    -------
+    tuple
+        fig, ax = (matplotlib.figure.Figure, fig.add_subplot)
+    """
+    fig = Figure(figsize=figsize)
+    ax = fig.add_subplot(projection=projection, **kwargs)
+    return fig, ax
+
+
+def default_figsize(figsize=(6.4, 4.8)):
+    """
+    Parameters
+    ----------
+    figsize : tuple
+            default (6.4, 4.8), same as mpl.rcParams["figure.figsize"]
+            but avoids global state
+    Returns
+    -------
+    tuple of sizes for figure
+
+    """
+    return figsize
```

Comment:
Hi @scarlehoff , @AleCandido . Thank you both for the comments.
I added this simple function. However, not sure whether this is the best way. A hard coded variable `default_figsize = (6.4,4.8)` would probably do as well.

Reply:
> A hard coded variable default_figsize = (6.4,4.8) would probably do as well.

Yes. It also helps (having a `DEFAULT_FIGSIZE = whatever`) for having a place where to change our default everywhere.

That said, I guess the problem with global states is when something gets modified. Reading constants like `mpl.rcParams` I would naively guess it is fine.

Reply:
> That said, I guess the problem with global states is when something gets modified. Reading constants like `mpl.rcParams` I would naively guess it is fine.

Definitely, but `mpl.rcParams` is also aware of user config files, potentially the distribution default configurations and so on.

If we have a preferred size, then having written somewhere is not a bad idea.

As @scarlehoff pointed out, they are usually written uppercase, when they are located top-level in a module, and not supposed to be modified.
For this reason they are often called *constants*, even if Python has no proper constant at all (even if it is top-level and upper case, you can always modify it, even though you shouldn't, most likely).

Reply:
Why is this needed? `Reportengine` takes a `default_style` argument which is set in vp (our settings are in `small.mplstyle`), does that not suffice?

Reply:
I completely forgot about those files. Apologies for that.

You're right, the default should be there.

Reply:
I'm not entirely sure what you mean by "would be possible to use the file content for direct usage?".

What currently happens is that vp (through reportengine) overwrites the global state of matplotlib using the `.mplstyle` file, i.e. it overwrites `rcParams`. This gives us the default settings for our plots. Of course, these can be read and overwritten locally if deemed necessary but based of the name of this function it seems you want to set another default which is why I asked.

Reply:
`rcParams` can be modified by `reportengine`, but also by other factors, since it is coming through `matplotlib`.

So, since a default style is set, it would be less surprising if all the values would come from there, rather than potentially pass through overridable matplotlib settings.
However, if the goal is to make it overridable, and just be consistent in the use of overridable properties, then reading `rcParams` is the correct course of action.